### ISSUE TITLE
`--disable-gil` builds are not compatible with the limited API

### ISSUE THREAD

**Issue by colesbury** [python/cpython#111506]
# Bug report

After https://github.com/python/cpython/pull/110764, `--disable-gil` builds of CPython 3.13 are (temporarily) not be compatible with the limited API or the stable ABI due to modifications to the `PyObject` header. For context, this was noted in the ["Backwards Compatibility"](https://peps.python.org/pep-0703/#backwards-compatibility) section of PEP 703.

Based on feedback on the related [discuss post](https://discuss.python.org/t/python-abis-and-pep-703/34018/) and discussions with Gregory Smith, Guido, Petr, and Victor at the core dev sprint, I believe we can make changes to the stable ABI so that extension author can target a single stable ABI that supports CPython 3.7+ and both the default build and the `--disable-gil` builds of CPython 3.13+. 

**These changes will require a PEP**. This issue is intended just to keep track of the known problem.

The rough idea is:
1) make a number of macro/inline functions (e.g., `Py_SIZE`, `Py_TYPE`, `Py_SET_REFCNT`, `Py_REFCNT`) "real" function calls in the stable ABI for CPython 3.13+
2) When targeting CPython <= 3.12, use dynamic symbol lookup (`GetProcAddress` on Windows) or weak symbols (on Linux/macOS) to determine at runtime if the active CPython version provides the desired functions. Otherwise, fall back to old definitions of the above functions for CPython <= 3.12.

### Linked PRs
* gh-111508
* gh-112747
* gh-112794
* gh-112798
* gh-117762
* gh-136505

**Comment by corona10** [comment]
cc @vstinner

**Comment by encukou** [comment]
**These changes will require a PEP.**

Please, could we have the PEP **first**, then do the changes?

**Comment by encukou** [comment]
> Please, could we have the PEP first, then do the changes?

@colesbury, do you still plan to write the PEP? Do you need any help?

**Comment by colesbury** [comment]
@encukou, yes I still intend to write the PEP. It just hasn't been as a high a priority as integrating the pieces of PEP 703.

**Comment by neutrinoceros** [comment]
Sorry if this was answered elsewhere already but I couldn't find this information: is the intention to bring some form of free-threaded limited API in CPython 3.13.0 final or later, in some subsequent minor version (3.14 at the earliest) ?

**Comment by danijar** [comment]
I'm curious if this still has a chance to make it into CPython 3.14?

My understanding compatibility with the limited API is basically blocking the extension ecosystem from preparing itself for a future where free threading gets enabled in the default CPython build. That's still some time away, but in parallel the ecosystem also needs time to evolve to allow that.

Right now, there is little incentive for smaller extension maintainers to go through the effort of building two wheels, because free threading requires a separate CPython build and thus isn't widely used yet. At the same time, enabling free threading in the main CPython build is infeasible until the extension ecosystem supports it widely (at least to the point where most extensions load, even if they re-enable the GIL on import).

The only way out of that I see is that the limited API becomes compatible with free threading as early as possible, allowing extension maintainers to support the new ABI at basically no cost, so that the ecosystem will be ready by the time free threading is stable and performant. Or am I misunderstanding the situation?

**Comment by clin1234** [comment]
Any update so far?

**Comment by ngoldbaum** [comment]
> Any update so far?

@encukou started a Discourse thread: https://discuss.python.org/t/making-pyobject-opaque-in-the-limited-api
